### PR TITLE
Extend Cosmos3 CPU offloading to component and layerwise paths

### PR DESCRIPTION
**Pull request by rahul-steiger-nv**
Extends CPU offloading support for Cosmos3.

## Purpose

Cosmos3 does not follow the usual separate text-encoder/DiT pipeline structure used by generic model-level CPU offload. Its transformer has a reasoner path that runs once and a generator path that runs during denoising.

This PR adds component-level model offloading for Cosmos3 by introducing a generic `ModelCPUOffloadMixin` and `GroupOffloadManager`, then using them to swap the reasoner/generator components inside the transformer forward pass.

It also extends Cosmos3 layerwise offload by marking the reasoner and generator block containers independently, and updates docs/tests for the new offload path.

The component offload manager is intentionally structured as an extension point. A follow-up PR will add an optimized weight-staging backend on top of this manager once the base Cosmos3 component-offload path is merged.

## Test Plan

**vLLM Version:** 0.23.0

**vLLM-Omni Commit:** 591f4d40

Focused unit tests:

```bash
VLLM_TARGET_DEVICE=empty /home/scratch.rsteiger_other/vllm-omni/.venv/bin/python -m pytest \
  tests/diffusion/models/cosmos3/test_cosmos3_transformer.py \
  tests/diffusion/models/cosmos3/test_cosmos3_pipeline.py \
  tests/diffusion/offloader/test_sequential_backend.py \
  -q
```

Benchmark setup:

- guardrails disabled with `--no-guardrails`
- model weights loaded offline from Hugging Face cache with `HF_HUB_OFFLINE=1`
- warmup=1, iters=3
- GPU memory cross-checked with `pynvml`

## Test Result

Focused unit tests:

```text
55 passed, 4 skipped
```

Cosmos3-Nano text-to-image on RTX 5090, 1024x1024, 50 steps:

| Mode | Median latency | Worker peak | Device peak | Result |
|---|---:|---:|---:|---|
| none | - | - | - | OOM |
| model offload | 4.18s | 21.97 GiB | 23.18 GiB | ok |
| layerwise offload | 14.68s | 9.88 GiB | 11.17 GiB | ok |

Cosmos3-Super text-to-image on RTX PRO 6000 Blackwell Server Edition, 1024x1024, 50 steps:

| Mode | Median latency | Worker peak | Device peak | Result |
|---|---:|---:|---:|---|
| none | - | - | - | failed during initialization |
| model offload | 15.19s | 67.38 GiB | 68.99 GiB | ok |
| layerwise offload | 56.62s | 11.88 GiB | 13.54 GiB | ok |

Cosmos3-Super text-to-video on RTX PRO 6000 Blackwell Server Edition, 1280x704, 89 frames, 35 steps:

| Mode | Median latency | Worker peak | Device peak | Result |
|---|---:|---:|---:|---|
| none | - | - | - | failed during initialization |
| model offload | 229.35s | 77.48 GiB | 79.09 GiB | ok |
| layerwise offload | 226.23s | 23.13 GiB | 24.79 GiB | ok |

These results show the expected tradeoff: for image generation, model-level component offload is much faster when it fits, while layerwise offload substantially reduces peak memory. For longer video workloads, layerwise offload can become the better tradeoff because the longer compute window gives weight transfers more opportunity to overlap with denoising compute, while peak memory is usually the limiting factor.

I did not include RTX 5090 video numbers because those runs would take too long in my local setup. The same reasoner/generator component-offload path is used for video, so the workload should still be supported when there is enough time and memory headroom.

### CODE REVIEW

**Review by hsliuustc0106 — COMMENTED** [review]
Cosmos3 component offloading design looks solid. Good documentation and benchmark coverage.

File: vllm_omni/diffusion/offloader/group_offload.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        if not self.enabled:
+            return
+        if name not in self.groups:
+            raise ValueError(f"Unknown offload group: {name!r} (known: {list(self.groups)})")
+        if self.active_group == name:
+            return
+        for other in self.groups:
+            if other != name:
+                self._offload_group(other)
+        self._load_group(name)
```
Comment (david6666666) [inline vllm_omni/diffusion/offloader/group_offload.py]: Can we clear `active_group` before the swap? If `_load_group(name)` fails after the previous group was offloaded, the manager still thinks the old group is active and a retry can skip reloading it.

File: tests/diffusion/models/cosmos3/test_cosmos3_transformer.py [hunk 2]
```diff
@@ -131,6 +132,59 @@ def test_forward_returns_video_prediction(monkeypatch: pytest.MonkeyPatch) -> No
...
+        def forward(self, hidden: torch.Tensor, **kwargs):
+            del kwargs
+            return hidden + self.weight.to(hidden.dtype)
+
+    model = transformer_cosmos3.Cosmos3VFMTransformer(
+        SimpleNamespace(tf_model_config=_tiny_cosmos3_config(), dtype=torch.float32)
+    )
+    model.language_model.layers = nn.ModuleList([ToyReasonerLayer()])
+    model.gen_layers = nn.ModuleList([ToyGeneratorLayer()])
+    model.enable_model_cpu_offload(device=torch.device("cpu"), pin_memory=False)
```
Comment (david6666666) [inline tests/diffusion/models/cosmos3/test_cosmos3_transformer.py:174]: This only exercises CPU->CPU, so `_to_cpu/_to_gpu` return early. Can we add a tiny accelerator-gated case that checks the reasoner/generator params actually move between CPU and the device?

File: vllm_omni/diffusion/offloader/group_offload.py (hunk 1 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/offloader/group_offload.py]: Done

File: tests/diffusion/models/cosmos3/test_cosmos3_transformer.py (hunk 2 above)
Comment (rahul-steiger-nv) [reply tests/diffusion/models/cosmos3/test_cosmos3_transformer.py:174]: Added it

**Comment by hsliuustc0106** [comment]
@yuanheng-zhao do you think it's necessary to introduce the group offloader?

**Comment by yuanheng-zhao** [comment]
> @yuanheng-zhao do you think it's necessary to introduce the group offloader?

As we seem to support single model as a special case for now, I'd recommend to support it in a model-local way if applicable, so that prevent adding more new abstractions at this moment. Maybe we could conclude requirements and design when we have multiple component-level models. 

While the direction of this PR (component-offload capability) is expected and nice, as the previous sequential offloading assumes common pipeline structures and swap between encoders/transformers.

cc @hsliuustc0106

**Comment by rahul-steiger-nv** [comment]
> As we seem to support single model as a special case for now, I'd recommend to support it in a model-local way if applicable, so that prevent adding more new abstractions at this moment. Maybe we could conclude requirements and design when we have multiple component-level models.

@hsliuustc0106 @yuanheng-zhao I removed the group offloader and added local model level support.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 3]
```diff
@@ -627,6 +633,10 @@ class Cosmos3OmniDiffusersPipeline(
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:677]: This breaks HSDP for Cosmos3. `_load_model_with_hsdp()` applies `apply_hsdp_to_model()` to every discovered DiT, and `Cosmos3LanguageModel` does not define `_hsdp_shard_conditions`, so HSDP will hit `transformer.language_model` first and raise before sharding the parent transformer.

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 4]
```diff
@@ -197,12 +197,31 @@ class ModelLevelOffloadBackend(OffloadBackend):
...
+            custom_enable(
+                device=self.device,
+                pin_memory=self.config.pin_cpu_memory,
+                use_hsdp=self.config.use_hsdp,
+            )
+            self._custom_pipeline = pipeline
+            self.enabled = True
+            logger.info(
+                "Model-level offloading enabled through %s.enable_omni_model_cpu_offload",
+                pipeline.__class__.__name__,
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/sequential_backend.py:225]: I'm thinking if we better make it a concise contract (i.e., `Protocol`), as for now the branch of enable_omni_model_cpu_offload and disable_omni_model_cpu_offload pairs are added with pure `getattr` detections. 

It should be good for now. We could update later

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 3 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:677]: @hsliuustc0106 Rebased, and updated `_load_model_with_hsdp` to shard only outermost DiTs via `PipelineModules.outermost_dits()` — a nested DiT (e.g. Cosmos3's `transformer.language_model`) is skipped since the ancestor's `_hsdp_shard_conditions` already cover its blocks. Tests added. Let me know if you'd prefer a different approach.

File: vllm_omni/diffusion/offloader/sequential_backend.py (hunk 4 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/offloader/sequential_backend.py:225]: I agree

**Comment by rahul-steiger-nv** [comment]
Hi! I've addressed the latest review comments and the CI is passing. When someone has time, I'd appreciate another review. Thanks!

File: vllm_omni/diffusion/offloader/sequential_backend.py [hunk 5]
```diff
@@ -259,6 +278,15 @@ def disable(self) -> None:
         if not self.enabled:
             return
 
+        if self._custom_pipeline is not None:
+            custom_disable = getattr(self._custom_pipeline, "disable_omni_model_cpu_offload", None)
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/sequential_backend.py:286]: Shall we leave a note or todo, as we discussed in https://github.com/vllm-project/vllm-omni/pull/4695/changes#r3497198161

File: vllm_omni/diffusion/offloader/module_collector.py [hunk 6]
```diff
@@ -22,6 +22,28 @@ class PipelineModules:
...
+        ``_hsdp_shard_conditions``; an inner DiT's blocks are already covered by
+        its ancestor's conditions, so sharding it again would double-wrap them
+        (and the inner stack may not declare conditions at all). Keeping only the
+        outermost DiTs leaves the ancestor's conditions as the single source of
+        truth for every nested block. Order follows ``dit_names``.
+        """
+        outer_names: list[str] = []
+        outer_modules: list[nn.Module] = []
+        for name, dit in zip(self.dit_names, self.dits):
+            nested = any(other is not dit and any(module is dit for module in other.modules()) for other in self.dits)
```
Comment (yuanheng-zhao) [inline vllm_omni/diffusion/offloader/module_collector.py]: nit: Shall we format and write it more explicit

**Review by yuanheng-zhao — COMMENTED** [review]
Thanks for revising the PR @rahul-steiger-nv , wonder if running the model with both HSDP and offloading enabled has been validated

File: vllm_omni/diffusion/offloader/sequential_backend.py (hunk 5 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/offloader/sequential_backend.py:286]: added

File: vllm_omni/diffusion/offloader/module_collector.py (hunk 6 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/offloader/module_collector.py]: done

**Comment by rahul-steiger-nv** [comment]
> Thanks for revising the PR @rahul-steiger-nv , wonder if running the model with both HSDP and offloading enabled has been validated

@yuanheng-zhao Thanks for the review.  

Validated on 2xGB200 (`Cosmos3-Nano`): `--use-hsdp` `--hsdp-shard-size 2 `shards only the outer `FSDPCosmos3VFMTransformer` (nested `language_model` not sharded independently) and composes cleanly with `--enable-cpu-offload`; output matches the single-GPU baseline.

**Review by yuanheng-zhao — APPROVED** [review]
LGTM, please resolve a trivial conflict @rahul-steiger-nv  . Thanks!

**Comment by rahul-steiger-nv** [comment]
> LGTM, please resolve a trivial conflict @rahul-steiger-nv . Thanks!

Thank you @yuanheng-zhao. Is there anything that I can do on my end to start the buildkite pipeline?

**Comment by yuanheng-zhao** [comment]
No worries. I've added label to trigger CI
